sklearn/linear_model/sag_fast.pyx

```diff
@@ -97,84 +292,131 @@ def sag(SequentialDataset dataset,
     cdef np.ndarray[double, ndim=1] cumulative_sums_array = \
         np.empty(n_samples, dtype=np.double, order="c")
     cdef double* cumulative_sums = <double*> cumulative_sums_array.data
+
     # the index for the last time this feature was updated
     cdef np.ndarray[int, ndim=1] feature_hist_array = \
         np.zeros(n_features, dtype=np.int32, order="c")
     cdef int* feature_hist = <int*> feature_hist_array.data
+
     # the previous weights to use to compute stopping criteria
-    cdef np.ndarray[double, ndim=1] previous_weights_array = \
-        np.zeros(n_features, dtype=np.double, order="c")
+    cdef np.ndarray[double, ndim=2] previous_weights_array = \
+        np.zeros((n_features, n_classes), dtype=np.double, order="c")
     cdef double* previous_weights = <double*> previous_weights_array.data
 
+    cdef np.ndarray[double, ndim=1] prediction_array = \
+        np.zeros(n_classes, dtype=np.double, order="c")
+    cdef double* prediction = <double*> prediction_array.data
+
+    cdef np.ndarray[double, ndim=1] gradient_array = \
+        np.zeros(n_classes, dtype=np.double, order="c")
+    cdef double* gradient = <double*> gradient_array.data
+
     # the scalar used for multiplying z
     cdef double wscale = 1.0
 
-    # helpers
-    cdef double val, cum_sum, gradient_change
-    cdef int j
-
     # the cumulative sums for each iteration for the sparse implementation
     cumulative_sums[0] = 0.0
 
+    # Loss function to optimize
+    cdef LossFunction loss
+    # Wether the loss function is multinomial
+    cdef bint multinomial = False
+    # Multinomial loss function
+    cdef MultinomialLogLoss multiloss
+
+    if loss_function == "multinomial":
+        multinomial = True
+        multiloss = MultinomialLogLoss()
+    elif loss_function == "log":
+        loss = Log()
+    elif loss_function == "squared":
+        loss = SquaredLoss()
+    else:
+        raise ValueError("Invalid loss parameter: got %s instead of "
+                         "one of ('log', 'squared', 'multinomial')"
+                         % loss_function)
+
     with nogil:
         start_time = time(NULL)
         while True:
```

Comment:
You could do either that or change this to `while not infinity` to remove the `if infinity: break` blocks in between